frontend/src/service-worker.js

```diff
@@ -1,15 +1,89 @@
+import {ExpirationPlugin} from 'workbox-expiration';
+import {precacheAndRoute, createHandlerBoundToURL} from 'workbox-precaching';
+import {registerRoute} from 'workbox-routing';
+import {NetworkFirst, StaleWhileRevalidate} from 'workbox-strategies';
 
-addEventListener('fetch', fetchListener);
+clientsClaim();
 
-addEventListener('install', function (e) {
-  e.waitUntil(self.skipWaiting());
-});
+if (process.env.NODE_ENV === 'production') {
+  // Precache all of the assets generated by your build process.
+  // Their URLs are injected into the manifest variable below.
+  // This variable must be present somewhere in your service worker file,
+  // even if you decide not to use precaching. See https://cra.link/PWA
+  precacheAndRoute(self.__WB_MANIFEST);
+}
+
+// Set up App Shell-style routing, so that all navigation requests
+// are fulfilled with your index.html shell. Learn more at
+// https://developers.google.com/web/fundamentals/architecture/app-shell
+const fileExtensionRegexp = new RegExp('/[^/?]+\\.[^/]+$');
+registerRoute(
+  // Return false to exempt requests from being fulfilled by index.html.
+  ({request, url}) => {
+    // If this isn't a navigation, skip.
+    if (request.mode !== 'navigate') {
+      return false;
+    }
+
+    // If this is a URL that starts with /_, skip.
+    if (url.pathname.startsWith('/_')) {
+      return false;
+    } // If this looks like a URL for a resource, because it contains // a file extension, skip.
+
+    if (url.pathname.match(fileExtensionRegexp)) {
+      return false;
+    }
+
+    // Return true to signal that we want to use the handler.
+    return true;
+  },
+  createHandlerBoundToURL(process.env.PUBLIC_URL + '/index.html'),
+);
+
+if (process.env.NODE_ENV === 'production') {
+  registerRoute(
+    ({url}) => {
+      const urlString = url.toString();
+      if (
+        urlString.startsWith('https://cdn.jsdelivr.net/') // Pyodide
+        || urlString.startsWith('https://pyodide-cdn2.iodide.io') // Only used when we are testing bleeding-edge pyodide
+        || url.hostname.endsWith('futurecoder.io')
+        || url.hostname.includes("localhost")
+        || url.hostname.includes("127.0.0.1")
+      ) {
+        return true
+      }
+      return false;
+    },
+    new StaleWhileRevalidate({
+      cacheName: 'everything',
+      cacheExpiration: {
+        maxEntries: 50,
+      },
+      cacheableResponse: {statuses: [0, 200]},
+      plugins: [
+        // Ensure that once this runtime cache reaches a maximum size the
+        // least-recently used stuff is removed.
+        new ExpirationPlugin({maxEntries: 30}),
+      ],
+    }),
+  );
+}
 
-addEventListener('activate', function (e) {
-  e.waitUntil(self.clients.claim());
+// This allows the web app to trigger skipWaiting via
+// registration.waiting.postMessage({type: 'SKIP_WAITING'})
+self.addEventListener('message', (event) => {
+  if (event.data && event.data.type === 'SKIP_WAITING') {
```

Comment:
We're not sending any such message anywhere. Just replace this with `skipWaiting` from `workbox-core`.

Reply:
So just replacing with 

```javascript
self.skipWaiting()
```

right?

(What was there is just what was from the `cra-template-pwa` example)

Reply:
> self.skipWaiting()

No, import skipWaiting from workbox-core, like clientsClaim.

> (What was there is just what was from the cra-template-pwa example)

I know, just cleaning up, and I do want to call skipWaiting eagerly for the non-caching stuff.

Reply:
Tested and it seems to work the same if it's just `self.skipWaiting()` eagerly, so we'll go with that.

Reply:
No, use the workbox function or copy the source code so that it's inside the install/activate event listener. That's what's always suggested and I'd like to stick to that unless I know why it's usually like that and why it wouldn't apply to us. That's still sufficiently eager.